Feature/3238/lambda error messages

agduncan94 opened the pull request:
https://github.com/dockstore/dockstore/issues/3238

* Tracks lambda events (push, delete and install) and stores into the database (new table)
* Adds a new endpoint for grabbing all lambda events for a user
* Change some existing endpoints to use open API
* Adds new endpoint for adding a GitHub installation event to Dockstore (called by lambda)

~The Postgres issue (unrelated) is failing the Travis ci.~

Currently, I just added the one endpoint for fetching all events for a user, though this may need changes or maybe even a new endpoint based on what people think is necessary.

For discussion:
One situation that I can think of that has a bit of an issue. If a user not on Dockstore tries to create/update a workflow that has the GitHub app installed (by pushing a branch with a dockstore.yml), this will throw an error since we can't associate it with a user. The issue is, who do we show this event to? We have no user associated, so the get user events doesn't work.

denis-yuen commented:
> One situation that I can think of that has a bit of an issue. If a user not on Dockstore tries to create/update a workflow that has the GitHub app installed (by pushing a branch with a dockstore.yml), this will throw an error since we can't associate it with a user. The issue is, who do we show this event to? We have no user associated, so the get user events doesn't work.

Could be reasonable to throw these away (or rather, just have them logged in AWS for our purposes but not shown to anyone)

denis-yuen commented on dockstore-webservice/src/main/java/io/dockstore/webservice/core/LambdaEvent.java:
```diff
@@ -0,0 +1,175 @@
+    @ApiModelProperty(value = "The type of event.", required = true, position = 7)
+    private LambdaEventType type;
+
+    @ManyToOne
```
FYI, defaults to eager

denis-yuen commented on dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java:
```diff
@@ -267,14 +271,21 @@ private WorkflowVersion updateDBVersionSourceFilesWithRemoteVersionSourceFiles(W
+            lambdaEvent.setMessage(msg);
+            lambdaEvent.setSuccess(false);
+            lambdaEventDAO.create(lambdaEvent);
+            sessionFactory.getCurrentSession().getTransaction().commit();
```
A little wary of this in production code, why is this necessary and has this pattern been used before?

agduncan94 replied:
Since an error is being thrown, the transaction is rolled back, so the message would never be written. We have used this once before for Zenodo I believe.

agduncan94 replied:
What if I clear the session before updating(sessionFactory.getCurrentSession().clear()). From what I understand this would mean that no unexpected things would be saved, only the lambda event.

agduncan94 replied:
The docs say that a clear 

> ... cancel all pending saves, updates and deletions.

denis-yuen replied:
That does sound a bit safer

denis-yuen commented on dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java:
```diff
@@ -317,19 +332,48 @@ private WorkflowVersion updateDBVersionSourceFilesWithRemoteVersionSourceFiles(W
+            lambdaEvent.setSuccess(false);
+            lambdaEvent.setMessage(ex.getMessage());
+            lambdaEventDAO.create(lambdaEvent);
+            sessionFactory.getCurrentSession().getTransaction().commit();
```
Ditto

denis-yuen commented on dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java:
```diff
@@ -1935,14 +1936,41 @@ public void deleteWorkflow(@ApiParam(hidden = true) @Parameter(hidden = true, na
+    @Consumes(MediaType.APPLICATION_FORM_URLENCODED)
+    @UnitOfWork
+    @RolesAllowed({ "curator", "admin" })
+    @Operation(description = "Handle the installation of our GitHub app onto a repository or organization.", security = @SecurityRequirement(name = OPENAPI_JWT_SECURITY_DEFINITION_NAME), responses = @ApiResponse(responseCode = "418", description = "This code tells AWS Lambda not to retry."))
```
Might have lost context, why is there a new endpoint for this as opposed to the existing  release endpoint. 
Is this only for reporting lambda installs?

agduncan94 replied:
Ya, this is purely for reporting lambda installs. Users could technically see on GitHub which repos they have the app installed on, but thought they might want to know dockstore knows too.

denis-yuen commented on dockstore-webservice/src/main/java/io/dockstore/webservice/resources/UserResource.java:
```diff
@@ -762,6 +766,17 @@ public Limits setUserLimits(@ApiParam(hidden = true) @Auth User authUser,
+    @ApiOperation(value = "See OpenApi for details")
+    public List<LambdaEvent> getUserGitHubEvents(@ApiParam(hidden = true) @Parameter(hidden = true, name = "user", in = ParameterIn.HEADER) @Auth User authUser) {
+        final User user = userDAO.findById(authUser.getId());
+        return lambdaEventDAO.findByUser(user);
```
I realise this is just a draft, but this will quickly need pagination or something to cut things down. Plus sorting

agduncan94 replied:
Do you know of a good endpoint that has pagination?

denis-yuen replied:
Offset and limit control pagination here 
https://github.com/dockstore/dockstore/blob/develop/dockstore-webservice/src/main/java/io/dockstore/webservice/jdbi/EntryDAO.java#L176
Looks like its used by the old tools and workflows pages that paginate

coverbeck commented:
> For discussion:
> One situation that I can think of that has a bit of an issue. If a user not on Dockstore tries to create/update a workflow that has the GitHub app installed (by pushing a branch with a dockstore.yml), this will throw an error since we can't associate it with a user. The issue is, who do we show this event to? We have no user associated, so the get user events doesn't work.

Not sure I like this proposal, but should the event go to all of the workflow's users? For example, if you and I are both collaborating on a workflow, and I break the .dockstore.yml, wouldn't you want to know? Sort of analogous to CI systems, where everybody who's pushed to the branch gets notified with failures.

coverbeck commented on dockstore-webservice/src/main/java/io/dockstore/webservice/core/LambdaEvent.java:
```diff
@@ -0,0 +1,175 @@
+
+    @Column(columnDefinition = "TEXT")
+    @ApiModelProperty(value = "The name of the user on GitHub that triggers the event.", required = true, position = 2)
+    private String username;
```
I think the name of this should be something like `gitHubUsername`. I at first started to put comments about it thinking it was the Dockstore username, until I read you annotations.

coverbeck commented on dockstore-webservice/src/main/java/io/dockstore/webservice/core/LambdaEvent.java:
```diff
@@ -0,0 +1,175 @@
+    private User user;
+
+    @Column(columnDefinition = "TEXT")
+    @ApiModelProperty(value = "The installation ID of the event.", position = 9)
```
Isn't this semi-secret? Probably not, but just double-checking.

agduncan94 replied:
Not entirely sure, will investigate.

agduncan94 replied:
After investigating I am not entirely sure, though I think it is ok just to remove it since it is not a requirement. We don't do anything with it, so I am just going to remove it.

coverbeck commented on dockstore-webservice/src/main/java/io/dockstore/webservice/core/LambdaEvent.java:
```diff
@@ -0,0 +1,175 @@
+    @JoinColumn(name = "userId", referencedColumnName = "id")
+    @ApiModelProperty(value = "User that the event is acting on (if exists in Dockstore).", position = 8)
+    @JsonIgnore
+    private User user;
```
Should this be lazy?

agduncan94 replied:
Yes, it is now, just forgot to push changes.

agduncan94 commented:
Discussion with Denis:
Find events by github org instead of by user. On the my workflows page, per org, there is a button to open modal which has lambda events for that org.

Might also make sense to have a per user one on the account page.

agduncan94 commented on dockstore-webservice/src/main/java/io/dockstore/webservice/core/LambdaEvent.java:
```diff
@@ -0,0 +1,175 @@
+    private String message;
+
+    @Column(nullable = false, columnDefinition = "boolean default false")
+    @ApiModelProperty(value = "Whether or not the user has dismissed the event.", position = 6)
```
Based on a discussion with Denis, this will only track the user who triggered the event. For other users to be able to see the events, we will use an endpoint which if a user has access, will show all events for a GitHub organisation. This way we don't need to keep track of all the users.

denis-yuen replied:
Following on, if all users in the org can see the event, it seems that we won't be able to track it with one boolean (I'm ok with not tracking this)

Ldcabansay replied:
Depending on how intrusive the UI will be, I think its fair not to dismiss the event or track which users have seen it. The error should disappear when a new change is pushed (and hopefully fixed). We don't track or dismiss the validation error messages on the files tab for example.

agduncan94 commented:
I've added another endpoint for grabbing lambda events for an organisation, added pagination, and added some tests. Also I removed github installation id.

agduncan94 commented:
@denis-yuen Good point, should I remove that attribute?

agduncan94 commented on dockstore-webservice/src/main/java/io/dockstore/webservice/core/LambdaEvent.java:
```diff
@@ -0,0 +1,165 @@
+        @NamedQuery(name = "io.dockstore.webservice.core.LambdaEvent.findByUser", query = "SELECT lambdaEvent FROM LambdaEvent lambdaEvent WHERE lambdaEvent.user = :user"),
+})
+@SuppressWarnings("checkstyle:magicnumber")
+public class LambdaEvent {
```
I think I am ok just renaming later if the need ever comes up.

agduncan94 commented on dockstore-webservice/src/main/java/io/dockstore/webservice/resources/LambdaEventResource.java:
```diff
@@ -0,0 +1,74 @@
+            @ApiParam(value = PAGINATION_OFFSET_TEXT) @QueryParam("offset") @DefaultValue("0") String offset,
+            @ApiParam(value = PAGINATION_LIMIT_TEXT, allowableValues = "range[1,100]", defaultValue = PAGINATION_LIMIT) @DefaultValue(PAGINATION_LIMIT) @QueryParam("limit") Integer limit) {
+        // To ensure a user has access to an organization, check that they have at least one workflow from that organization
+        List<Workflow> workflows = workflowDAO.findMyEntries(user.getId());
```
I like this idea, I'll try one of the approaches.

denis-yuen commented on dockstore-webservice/src/main/java/io/dockstore/webservice/resources/LambdaEventResource.java:
```diff
@@ -62,11 +65,12 @@ public LambdaEventResource(SessionFactory sessionFactory) {
-        List<Workflow> workflows = workflowDAO.findMyEntries(user.getId());
-        boolean canAccessOrganization = workflows.stream().anyMatch(workflow -> Objects.equals(workflow.getOrganization(), organization) && Objects.equals(workflow.getSourceControl(),
+        User authUser = userDAO.findById(user.getId());
+        List<Workflow> workflows = workflowDAO.findByOrganization(organization);
```
This cuts down data loaded in memory from all workflows to an organization's workflows, which is a lot better. 
But can still be better by avoiding loading workflows at all since checking that the user has access can be done purely by query or checking with GitHub that the user had access to the organization directly in the case that the user has not been added to a workflow in the org yet would be more efficient. 
Can be added as a code comment (TODO) and ticket

agduncan94 replied:
I ended up just changing it to call the GitHub API, much cleaner now.

Ldcabansay commented on dockstore-webservice/src/main/java/io/dockstore/webservice/resources/ResourceConstants.java:
```diff
@@ -33,7 +33,13 @@
     public static final String NIHDATACOMMONS = "Needed for SmartAPI compatibility apparantly, might be cargo cult behaviour";
+    public static final String LAMBDAEVENTS = "Query lambda events triggered by GitHub Apps";
     public static final String OPENAPI_JWT_SECURITY_DEFINITION_NAME = "bearer";
+    public static final String PAGINATION_LIMIT = "100";
```
This seems like a high limit to me. Can always adjust later though.
